File: ios/HybridCameraVisionPixelColors.swift
//
//  HybridCameraVisionPixelColors.swift
//  CameraVisionPixelColors
//

import Foundation
import NitroModules

class HybridCameraVisionPixelColors: HybridCameraVisionPixelColorsSpec {
  func analyzeImageAsync(image: ImageData) throws -> Promise<PixelColorsResult> {
    return Promise { resolve, reject in
      DispatchQueue.global(qos: .userInitiated).async {
        let width = Int(image.width)
        let height = Int(image.height)
        let data = Data(image.data)

        let result = PixelAnalyzerEngine.shared.analyzeImageData(
          width: width,
          height: height,
          data: data
        )

        let uniqueColorCount = result["uniqueColorCount"] as? Int ?? 0
        let topColorsDict = result["topColors"] as? [[String: Int]] ?? []
        let brightestColorsDict = result["brightestColors"] as? [[String: Int]] ?? []

        let topColors = topColorsDict.map { dict in
          RGBColor(
            r: Double(dict["r"] ?? 0),
            g: Double(dict["g"] ?? 0),
            b: Double(dict["b"] ?? 0)
          )
        }

        let brightestColors = brightestColorsDict.map { dict in
          RGBColor(
            r: Double(dict["r"] ?? 0),
            g: Double(dict["g"] ?? 0),
            b: Double(dict["b"] ?? 0)
          )
        }

        let pixelColorsResult = PixelColorsResult(
          uniqueColorCount: Double(uniqueColorCount),
          topColors: topColors,
          brightestColors: brightestColors
        )

        resolve(pixelColorsResult)
      }
    }
  }
}

File: android/src/main/java/com/margelo/nitro/cameravisionpixelcolors/HybridCameraVisionPixelColors.kt
package com.margelo.nitro.cameravisionpixelcolors

import com.cameravisionpixelcolors.PixelAnalyzerEngine
import com.margelo.nitro.core.Promise
import java.util.concurrent.Executors

class HybridCameraVisionPixelColors: HybridCameraVisionPixelColorsSpec() {
    private val executor = Executors.newSingleThreadExecutor()

    override fun analyzeImageAsync(image: ImageData): Promise<PixelColorsResult> {
        return Promise.async {
            val width = image.width.toInt()
            val height = image.height.toInt()
            val data = ByteArray(image.data.size.toInt())
            image.data.getBuffer(false).get(data)

            val result = PixelAnalyzerEngine.analyzeImageData(width, height, data)

            val uniqueColorCount = result["uniqueColorCount"] as? Int ?: 0
            @Suppress("UNCHECKED_CAST")
            val topColorsMap = result["topColors"] as? List<Map<String, Int>> ?: emptyList()
            @Suppress("UNCHECKED_CAST")
            val brightestColorsMap = result["brightestColors"] as? List<Map<String, Int>> ?: emptyList()

            val topColors = topColorsMap.map { map ->
                RGBColor(
                    r = (map["r"] ?: 0).toDouble(),
                    g = (map["g"] ?: 0).toDouble(),
                    b = (map["b"] ?: 0).toDouble()
                )
            }.toTypedArray()

            val brightestColors = brightestColorsMap.map { map ->
                RGBColor(
                    r = (map["r"] ?: 0).toDouble(),
                    g = (map["g"] ?: 0).toDouble(),
                    b = (map["b"] ?: 0).toDouble()
                )
            }.toTypedArray()

            PixelColorsResult(
                uniqueColorCount = uniqueColorCount.toDouble(),
                topColors = topColors,
                brightestColors = brightestColors,
                motion = null,
                roiApplied = null
            )
        }
    }
}
